[RunAllTests] Fix part of #5343: Upload generated code coverage report as comments

Rd4dev opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Fixes part of #5343 

### Project
[PR 2.4 of Project 4.1]

### Changes Made

With the protos collected and stored with their respective file paths through #5465, this PR will focus on collecting the stored proto files and passing it to the script to generate a coverage report along with a status check to decide the success failure case of the CI Coverage Check run.

**Collection of protos**
- The stored protos are now being uploaded as an artifact.
- The name of the artifact is set to a dynamic value corresponding to its shard_name.
  - The shard_name is extracted from the first portion of the `CHANGED_FILES_BUCKET_BASE64_ENCODED_SHARD` value.
    - eg: The matrix job - [app-shard0;H4sIAAAAAAAAAONiTiwoEHIFEvrFRcn6uYmZefpZi....] will save the artifact with name **coverage-report-app-shard0**
    - The dynamic or unique name is important to prevent the following error: 
```
Error: Failed to CreateArtifact: Received non-retryable error: Failed request: (409) Conflict: 
an artifact with this name already exists on the workflow run
```
- The artifacts that match the pattern of coverage-report-* are downloaded.

```
Found 4 artifact(s)
Filtering artifacts by pattern 'coverage-report-*'
Preparing to download the following artifacts:
- coverage-report-domain-shard0 (ID: 1799350592, Size: 489)
- coverage-report-scripts-shard2 (ID: 1799348996, Size: 2075)
- coverage-report-app-shard3 (ID: 1799348891, Size: 770)
- coverage-report-generic-shard1 (ID: 1799348719, Size: 551)
```
- All the stored coverage_report.pb files are found and their paths are stored as list.
- The list of coverage_report.pb files are then passed to the CoverageReporter.kt script to handle the coverage report protos.
- The script combines them as one single CoverageReportContainer to generate a Markdown report with it.
- Finally, a .md report is stored and based on the status of the coverage check, the job either succeeds or fails.
- The stored md report is again uploaded and downloaded as an artifact to pass it for publication of the report.
- The CoverageReport.md file is then uploaded as a comment to the corresponding PR using - https://github.com/peter-evans/create-or-update-comment action

>The comment will be published regardless of whether the coverage passes or fails. The only time comments will not be published is if the unit tests themselves fail, making the coverage checks non-functional (also less cluttered). 

#

### CI Run Data

This check was run with 4 different shards on all possible cases capable of being produced.
Check : https://github.com/oppia/oppia-android/actions/runs/10340867060?pr=5469
Comment Published with the above run: https://github.com/oppia/oppia-android/pull/5469#issuecomment-2282804068

The above cases are run with custom temporary changes to the test exemption file (testing purposes).
```
test_file_exemption {
  exempted_file_path: "scripts/src/java/org/oppia/android/scripts/testfile/TestFileCheck.kt"
  override_min_coverage_percent_required: 101
}
test_file_exemption {
  exempted_file_path: "utility/src/main/java/org/oppia/android/util/logging/ConsoleLogger.kt"
  override_min_coverage_percent_required: 30
}
```

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

Rd4dev committed:
Save lobby - Collecting protos into one coverage report container proto that can be uploaded later as an artifact and utilized later for one combined report generation

Rd4dev committed:
Updating the coverage missing test case while removing the job and replicating the actual failure scenerio by not actually using a test file for its testing purposes

Rd4dev committed:
Clean up and changed the test case names to be more specific on what causes their failure

Rd4dev committed:
Added validation to ensure exactly one unique value of filePath and fileSha1Hash are present in the coverage report corresponding to one file for n:1 targets included

Rd4dev committed:
Clean up of aggreagation code and replaced conditional error statments with kotlin checks

Rd4dev committed:
Removed the mock data used for coverage report with n:1 file target assumption

Rd4dev committed:
Resolve merge conflicts with changes to upstream one to may targets branch

Rd4dev committed:
Hardcoded implementation of saving the protos to their own path

Will need to take in a dynamic path

Rd4dev committed:
New proto structure for coverage to having distinct case for success details and failure message case in Coverage Report

Rd4dev committed:
Roughly organized layout / implementation of the flow of the protos and report generation

The base flow is RunCoverage runs CoverageRunner on each file and acquires a single or a list of protos based on the relationship with targets and test files. With the collected protos, it can either have a success (details) or a failure case. If any of them is a failure then it returns a failure for the file's test target, if the file has a success case then aggregation occurs to account incase of multi relationship cases. Till this each file will now have their own 1 single coverage report. And each of them are finally combined as one single container of coverage. This container is then passed to the CoverageReporter. Here the coverage report container is taken as one single unit and looped for each coverage report for both html and md. With html for each success case a html file needs to be created. With md report both success and failure cases should be taken and included in the final report as just one single report in one common spot (this is included just to maintain the standard flow other than that and if needed debugging, the md woudln't serve anything useful for local dev works)

Rd4dev committed:
Re-writing the CoverageReporter script to handle the container proto and success, failure and exemption cases specifically for HTML and MD reports. Till this commit the log and html sections are re written while md report and proto storage needs to be handled

Update tests later :(

Rd4dev committed:
First set of cleanup with coverage reporter implemented

Pending status check addition, exemption list addition, saving protos

Rd4dev committed:
Reference point for mock data with new repeated bazel test target proto structure

Rd4dev committed:
Clean up and simplification of usage of single usage entry

Rd4dev committed:
Fix lint checks and failing test case with update to the proto strucuture

Rd4dev committed:
Added missing import for BazelTestTarget

Rd4dev committed:
Final md report generation with failure, details - 1. failure below threshold 2. exempted failure 3. success 4. exempted success and test file exemption cases

This just feels like a dumped version would require major refactorings and also does have lot of testing data that needs to be cleaned up

Rd4dev committed:
Refactored CoverageReporter into unit functionalities and avoid repetitions

The md report will be added as part of their respective success and failure tables with appropriate percentages requires and felt nice to be at one place than at a different spot or separte section

Rd4dev committed:
Added parameter to save the proto to the path specified

This though wouldn't be useful for local dev but is much needed to save the protos of each matrix run. This could be saved with their encoded bucket names and retrieved later

Rd4dev committed:
Added Coverage Check Status that check through every fail and details case and decides on a Coverage Status value as either PASS or FAIL

This will later be passed to CI to run the final Run Coverage Status Check that can fail the entire workflow if FAIL status is retrieved and only pass on PASS

Rd4dev committed:
Added a helper method to calculate coverage percentages

Rd4dev committed:
Updated tests for Coverage Runner Test

Rd4dev committed:
Added test case for CoverageReporter to test success detail, falure and exemption cases with html report generation

Rd4dev committed:
Added test case for CoverageReporter to test success detail, falure and exemption cases with md report generation

Rd4dev committed:
Added combined success failure and exempted test file cases

Rd4dev committed:
Updated RunCoverageTest test cases with new rewritten report implementations

Rd4dev committed:
Resolve merge conflicts with changes to upstream proto BazelTestTarget, test additions

Rd4dev committed:
Fix lint checks indents and max line lengths

Rd4dev committed:
Fix test with unresolved bazel test target list, added kdoc

Rd4dev committed:
Fix failing test cases, lint checks indents, missing new lines and max line lengths

Todo: 1. The Coverage Status Checks needs to be utilized to have the check pass or fail with error and their respective tests needs an update accordingly. 2. The new lines needs to be properly handled with conditions and again update tests accordingly.

Rd4dev committed:
Fix failing test with absence of trimEnd

Rd4dev committed:
Conditional new line additions; with next commits replace them with append

Rd4dev committed:
Replaced appendLine with append to use \n in line

Rd4dev committed:
Fix Tests with Failure cases and used the public const MIN_THRESHOLD throughout the test and sources

Rd4dev committed:
Kdoc checks and fix test with list of files

Rd4dev committed:
Added Coverage Check Status to entire check and md report

Rd4dev committed:
Refactored new lines in tests and added Coverage status to test templates

Rd4dev committed:
Fix failing tests with color codes and new line missing

Rd4dev committed:
Updated test failure with appropriate failure type in test cases

Rd4dev committed:
Fix failing test cases with missing new line

Rd4dev committed:
Bumping shard_count to 24 that got overriden with merge

Rd4dev committed:
Added new tests with failing cases and new types of reports

Rd4dev committed:
Fix Lint check removed unused imports

Rd4dev committed:
Readding the shard count, removing while local testing

Rd4dev committed:
Refactored log report to log only the failed coverage report to avoid clutter

Rd4dev committed:
Fixed Kdoc issues and redundant quotes

Rd4dev committed:
Resolve merge conflict with upstream list of files and modified always with !cancelled in code coverage check

Rd4dev committed:
Added remaining test target collection test cases for retrieveChangeFilesTest

Rd4dev committed:
Introducing ComputeChangedFilesTest test file

Rd4dev committed:
Base test case setup to add a file as a changed file for computeChangedFilesTest case

ComputeAffectedTestsTest would require creating test files but with this this should be just files

Rd4dev committed:
Added test cases for ComputeChangedFiles replicated exactly from the ComputeAffectedTestsTest but with file changes

Still needs to enable the compute all files setting to have it run all files with Run All Tests and have the coverage workflow triggered only after unit tests workflow is done

Rd4dev committed:
Enable Functionality to compute on all available .kt source files

Rd4dev committed:
Merge branch 'develop' of https://github.com/Rd4dev/oppia-android into code_coverage_many_to_one_targets

Rd4dev committed:
Added the copy of local and shared test to ensure the multiple test for a file do work properly

Rd4dev committed:
Resolve merge conflict with upstream RunCoverageTest

Rd4dev committed:
Added test to verify that the final container is saved to output proto path specified

Rd4dev committed:
Save lobby for proto extraction and storage

Rd4dev committed:
Added test case to verify one single line hit my multiple targets of a single source file

Rd4dev committed:
Updated the Final coverage report template as suggested. This introduces having a table for the No of files, total coverage and coverage analysis, but the template suggestion was to have them as individual lines. Having this as a save point and if approved will be modified later

Rd4dev committed:
Updated tests with updated md template and fixed few review comments

Needs a lot of iterations with test cases and also with the review comment to have specific testing scenerios, that should be the way but would require re iterating again. And also have calls for test content cases to simply it. Also the report structure tests too are not completely tested locally will update with ci reports as they should be much faster

Rd4dev committed:
Reverting the shard count change for ci run

Rd4dev committed:
Resolve merge conflicts with upstream one to many targets :(

Rd4dev committed:
Save lobby while fixing tests as they take too long

With changes to test exemption take as input and changes to the markdown every test case needs and update and taking too long

Rd4dev committed:
Enabled and refactored compute changed files to have all files setting to true and get all files

Needs to be tested in ci and test related to compute all files need to be added

Rd4dev committed:
Remove unintended shard count

Rd4dev committed:
Cleaned up the BazelClient could be refactored more, will do after 1st review

Rd4dev committed:
Little clean with left debugging lines

Rd4dev committed:
Added test case with compute all file settings

Rd4dev committed:
Refactoring with test file exemption removal

Would probably need to fix failing test cases first to actually see coverage workflows in action, will switch to PR 2.2

Rd4dev committed:
Fixed test cases with Coverage Reporter and also added test cases corresponding to low and high pass fail exemption percentages

Rd4dev committed:
Fix failing test with CoverageRunnerTest due to incorrect target specification

Rd4dev committed:
Fixing test cases part 1 - md report template changes

Rd4dev committed:
Fix Lint checks majorly formattings

Rd4dev committed:
Fix tests part 2 - exceptions and templates

Rd4dev committed:
Fix Tests part 3 - remove unnecessary lines

Rd4dev committed:
Fix test after bumping min threshold to 70

Rd4dev committed:
Fix tests part 4 - with template and min threshold changes :|

Rd4dev committed:
Fix Lint check Remove needless blank line

Rd4dev committed:
Updating with the upstream list of files

All tests are finally updated and passing, with this the coverage runs should run successfully but only start after the unit tests pass

Rd4dev committed:
Removing the logic to run test with args

Rd4dev committed:
Bumping the processTimeout to 15 since the code coverage run failed with process did not finish within the time error was thrown

Rd4dev committed:
Fix Lint checks and re running ci to verify coverage pass

Rd4dev committed:
Removed the supress cases and comments

Rd4dev committed:
Bumping the process timeout straight in the RunCoverage to 35 and removing the need to wait for the unit test workflow to finish to see the results quickly

This is only done after verifying all the test have passed and its only the timeout issue with the ci runs

Rd4dev committed:
Trying to bump processTimeout in test

Rd4dev committed:
Brought in changes from the exempt-test-targets-incompatible-with-code-coverage branch

Rd4dev committed:
Changing the report format to proto for the ci runs and also bumping processTimeout in the ci runs

The format set to proto must introduce a check to see if proto output path exists to properly save it, and that will be re introduced with the upcoming pr when things make sense to determine where to store them

Rd4dev committed:
Resolved conflicts with upstream list of files

Rd4dev committed:
Fix Lint check missing space after =

Rd4dev committed:
Fix Lint check missing space after =

Rd4dev committed:
Triggering ci since many checks seem to fail after update

Rd4dev committed:
Remove old implementation for collecting data

Rd4dev committed:
Merge from upstream remove old implementation

Rd4dev committed:
Fix failing test cases with changes from pull from exemptions pr

Rd4dev committed:
Added tests for BazelClient and Missing Build declarations inclusions

Rd4dev committed:
Fix Lint checks max line legth

Rd4dev committed:
Merge with develop branch

Rd4dev committed:
Reordering the coverage output

Rd4dev committed:
Readjusted the bazel client test output

Rd4dev committed:
Merge with latest upstream list of files

Rd4dev committed:
Removed unused instrumentation / path finder variables

Rd4dev committed:
Updated md template with replacing the file name links with the details summary

Rd4dev committed:
Moved exempted list too to a details summary as they too might grow big

Rd4dev committed:
Updated tests to reflect changes in md report with replacement of links to details summary

Rd4dev committed:
Updated with suggested edits in the review

Rd4dev committed:
Removed parsing of the pb file in the Coverage Reporter Test

Rd4dev committed:
Fix failing tests and lint checks

Rd4dev committed:
Fixed a flaky test, expected it to be since the order depends on the bazel executions

Rd4dev committed:
Fix Lint check indentation

Rd4dev committed:
Merge from upstream list of files branch

Rd4dev committed:
Mapping tests to targets right in the compute changed files

This catches few things 1. Changed the source list returned earlier in the list of files pr to just return a string, earlier tried having .single() call on them as that could confirm it returning just one string path from the list, but turns out there are files that are not properly placed in the source code causing the mapping not appropriate so throwing an error would halt it, instead for those files a null is thrown and then filtered out. 2. Used distinct to just have one single source entry passed for the run coverage so no duplicate expensive runs 3. The mapping of test to source is now also a part of the compute changed files to make sure we only send sources to the run coverage and also having done this here would help with retrieving the appropriate test targets in the retrieve test targets for later builds. All these need to be tested with test cases and verified in the ci run

Rd4dev committed:
Code cleanup the changes need verification in ci

Rd4dev committed:
Removed steps: git-secret, decrypt secrets and all with caching steps since disabled for a long time

Rd4dev committed:
Split instrumentation test cases and added test for non kt files checks

Rd4dev committed:
Updated test cases and added tests for non kotlin files

Rd4dev committed:
Updating tests to work on files that are committed and not just added or staged or untracked, the script behaviour needs to be addressed in the upcoming commits

Rd4dev committed:
Fixed/Updated all the test cases with changes to the computing only the changed files

Rd4dev committed:
Reverting the commented check unit test completed job to verify ci runs

Rd4dev committed:
Resolve merge conflicts with all upstream list of files and ci workflows

This pr will need to do 3 main things 1. Make the coverage reporter as a runnable script with main call to run it while passing the pb file, 2. The pb file will be collection of pb file so a new script needs to be introduced to handle that with its appropriate test file 3. Each shard will need to upload its own proto file that needs to be somehow retrieved and combine. Ok reading it in the reverse order would make more sense but these are the key stuff that need to be done with this pr and finally uploading the report

Rd4dev committed:
Setup for having the coverage report for main call, moving around test exemption list to have a default value, so it can be skipped in the ci run

Rd4dev committed:
Added necessary changes to be accessible with cli - named as coverage_reporter

Rd4dev committed:
Logging: Still trying to figure out matrix workflows to collect pb files

Rd4dev committed:
Merge and resolving with upstream after incompatible files are exempted

Rd4dev committed:
Resolve conflict with coverage BUILD

Rd4dev committed:
Updated tests with source incompatible exemption changes

Rd4dev committed:
Fix Tests after merging with upstream exemptions

Needs a cleanup with handling of protos per test name paths

Rd4dev committed:
Storing the protos to their respective file name paths and updated its test cases

Rd4dev committed:
Fix tests with saving proto - yet to confirm test case with changes to coverage report collection type

Rd4dev committed:
Added log statement for md report path

Rd4dev committed:
Resolved merge conflict with upstream list of files

Rd4dev committed:
Commented the dependency to unit test to test coverage runs after caching is removed

Rd4dev committed:
Removing the need dependency for coverage workflow

Rd4dev committed:
Added test cases to cover the mapping of test to source file

Rd4dev committed:
Fix upstream override of findTestFile args

Rd4dev committed:
Fix failing test with additional message missing errors

Rd4dev committed:
Merge with upstream list of files after test fixes

Rd4dev committed:
Fix failing tests with high Coverage Exemption

Rd4dev committed:
Merge with upstream list of files

Rd4dev committed:
Reverting the commented out check unit job

Rd4dev committed:
Resolve merge conflict with upstream ci worflow branch

Rd4dev committed:
Removed older way of loading exemption

Rd4dev committed:
Uploading artifacts to collect the coverage reports dir

Rd4dev committed:
Reverting the single comma change with RunCoverage

As this is the only change reverting it and can be added later since having this would trigger coverage on it taking a light year to verify every ci run

Rd4dev committed:
Improper indentation causing invalid workflow

Rd4dev committed:
Removing the echo line

Rd4dev committed:
Changing the proto to MD as keeping the proto would skip the coverage status check as there is no call to the coverage reporter yet and that will only be introduced with the upcoming pr to upload the artifacts, and use them to derive a coverage status

Rd4dev committed:
Display all downloaded artifacts

Rd4dev committed:
Triggering ci with empty commit, the format here should be proto

Rd4dev committed:
Testing with proto insertion again as md was reported

Rd4dev committed:
Resolving conflict, guess this might have caused previous mis representation

Rd4dev committed:
Adding a new line change to app module file to create a new bucket and verify if the upload and downloads include them properly

Rd4dev committed:
Filtering out the pb files

The previous run was successful as it logged out both the app and the script pb reports, and downloading the app artifact did indeed have the desired proto data

Rd4dev committed:
Fix invalid workflow file

Rd4dev committed:
Passing the pb file list to the coverage reporter cli

Rd4dev committed:
Fix Invalid workflow file

Rd4dev committed:
Print repo root to avoid workflow fail

Rd4dev committed:
Save point while testing coverage report collection from pb files locally

The ci seems to need setup for bazel workspace to run scripts

Rd4dev committed:
Adding setup bazel actions

Rd4dev committed:
Fix invalid workflow due to incorrect indentation

Rd4dev committed:
Removing the matrix shard encoded name as they contain special characters and interfere with parsing the file

Rd4dev committed:
Adding in the cache and build environments

Rd4dev committed:
Added cache directory and checkout to have the codebase

Rd4dev committed:
Added step for cache for bazel

Rd4dev committed:
Uploading and adding comment job added

Rd4dev committed:
Upload needs to wait for report collection

Rd4dev committed:
Using body path instead of body

Rd4dev committed:
Removed the name for upload artifact

Rd4dev committed:
Add part of shard name

Rd4dev committed:
Having shard name need to figure out a way

Rd4dev committed:
Reading coverageReport.md and changing upload name

Rd4dev committed:
Changing downloadn name

Rd4dev committed:
Resolve conflict with develop

Rd4dev committed:
Min threshold conflict

Rd4dev committed:
Removing new line in RunCoverage don't want to run it as it takes too long

Rd4dev committed:
Resolve conflict

Rd4dev committed:
Changing the shard name structure withiout ;

Rd4dev committed:
Logging find as that seems to be malfuncitoning

Rd4dev committed:
Removed call before script

Rd4dev committed:
Pb file list and path prefix removed

Rd4dev committed:
Fixing the formats sent

Rd4dev committed:
pb file path saved to env

Rd4dev committed:
find collection and manipulation is the causer

Rd4dev committed:
Retrigger

Rd4dev committed:
Retrigger

Rd4dev committed:
Retriggering after removing commented liens

Rd4dev committed:
Including only one script runner job

Rd4dev committed:
save coverage report path to env

Rd4dev committed:
Non reachable null

Rd4dev committed:
Added test and coverage content as body

Rd4dev committed:
Fix unclosed warning

Rd4dev committed:
File path as path

Rd4dev committed:
commenting test due to ambiguity error

Rd4dev committed:
Workaround to atleast get comment out

Rd4dev committed:
Converting it to txt to display it

Rd4dev committed:
read text

Rd4dev committed:
add 2

Rd4dev committed:
Making changes to each type of fail, pass and exemption to see if the coverage report is properly reported

Rd4dev committed:
Artifact name conflicts saving to a common default directory

Rd4dev committed:
Changing the shard name to set to artifact name as it is conflicting, needs to be verified on ci

Rd4dev committed:
Reverting format lets see if we can use the existing ones

Rd4dev committed:
Typo causing failure

Rd4dev committed:
Shard name

Rd4dev committed:
Adding shard name to name ; lets see if this drops them in their individual folders

Rd4dev committed:
Retrigger since download fail

Rd4dev committed:
Adding condition to run it regardless of pass or fail status to always get the report

Rd4dev committed:
Saving them to a certain path while collecting them as a pattern

Rd4dev committed:
Invalid workflow file

Rd4dev committed:
Invalid workflow file

Rd4dev committed:
uncomment configure bazel

Rd4dev committed:
Adding run line

Rd4dev committed:
commenting to figure out the issue

Rd4dev committed:
ADding download

Rd4dev committed:
log dirs

Rd4dev committed:
Fixed the exemption template that missed the new line and added them conditionally based on the failure case presence and also adding another step just to try if we can directly get the md report from path instead of an extra .txt parsing step which again uploads and downlods the artifacts

Rd4dev committed:
Fix / update tests with changes

Rd4dev committed:
Clean up and temporarily exempting RunCoverage as that would fail and exempting it just to verify the next coverage run

Rd4dev committed:
Fix test with CoverageReporterTest color codes

Rd4dev committed:
Added docs and reverted all testing data

yet to fix the ambiguity error with main calls

Rd4dev committed:
Revert unintended removal

Rd4dev commented:
@BenHenning, the error on `overload resolution ambiguity:` is yet to be addressed, but can you PTAL on the implementations for uploading the comments? I have also added tests for the script modifications.

BenHenning commented on .github/workflows/code_coverage.yml:
```diff
@@ -253,3 +340,7 @@ jobs:
         if: ${{ needs.compute_changed_files.outputs.can_skip_files != 'true' && needs.code_coverage_run.result != 'success' }}
         run: exit 1
+
+      - name: Coverage status passed
```
```suggestion
      - name: Check that coverage status is passed
```

Rd4dev replied:
Done, updated as suggested.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/coverage/CoverageReporter.kt:
```diff
@@ -9,6 +9,63 @@ import java.io.File
 /** Minimum coverage percentage required. */
 const val MIN_THRESHOLD = 70
 
+/** ANSI escape codes for colors. */
```
```suggestion
/* ANSI escape codes for colors. */

```

This isn't a KDoc (since it has no element to document.

Rd4dev replied:
Done, updated as suggested.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/coverage/RunCoverage.kt:
```diff
@@ -15,21 +15,11 @@ import org.oppia.android.scripts.proto.TestFileExemptions
  * Entry point function for running coverage analysis for a source file.
  *
  * Usage:
- *    bazel run //scripts:run_coverage_for_test_target -- <path_to_root> <list_of_relative_path_to_files>
```
Nice catch.

Rd4dev replied:
Thanks :)

BenHenning reviewed (changes requested):
Thanks @Rd4dev! Just had a few comments for the base PR, but also some follow-up items:
1. It seems that the ``main()`` ambiguity issue is coming from adding the new ``main()`` function to ``CoverageReporter`` since ``RunCoverage`` depends on it here: https://github.com/oppia/oppia-android/blob/f1e3463f0ab4f39df20220efea7a2474fa663556/scripts/src/java/org/oppia/android/scripts/coverage/BUILD.bazel#L16
2. Please make sure to update this PR to account for the additional requests from the product review meeting yesterday (so that we can consider this project functionally complete after this PR is merged).
3. Once the ``main()`` is resolved, please update the PR title to run with ``[RunAllTests]`` so that we can verify coverage for all tests in the codebase. Now that you've verified coverage working for some targets (per the PR's comment history), we ought to keep 'RunAllTests' on for the remainder of the PR's life since we'll want to continually try all tests to make sure develop doesn't break once this PR is merged.
4. As part of (3) you'll probably need to add some additional test exemptions for tests that pass but have below 70% coverage (I'm positive there are at least a few from my own analysis). I suggest doing the following: finish the rest of the PR (including fixing the main issue) and waiting until CI finishes with a pass, then add 'RunAllTests' and re-run CI by closing and reopening the PR, once coverage runs without additional fixes being needed then update the PR to include any additional exemptions that are needed and verify CI fully passes.

Rd4dev committed:
Renamed step name for coverage check

Rd4dev committed:
Restructured the Coverage Reporter to its own package since ambiguity in finding the main call related to the corresponding script in test classes

Rd4dev committed:
Fix Lint checks and also a save point before reverting to a older commit to test main ambiguity workaround

Rd4dev committed:
Fix Lint check buildifier error

Rd4dev committed:
Fix Lint Checks and tests

I tried reverting to a certain commit, switching to a new branch and merging again, hopefully that doesn't do anything weird in the PR :|

Rd4dev committed:
Mergin with upstream upload-comments branch

Rd4dev committed:
Fix failing tests with new line changes adapted to fit conditional display

Rd4dev committed:
Added the reference note to the file exemptions list

The thing left to do is converting the list to a table as per demo suggestions

Rd4dev committed:
Updated the template for Exempted cases to now be in a tabular form and also added a new line to bunch of files belonging to differnt shards to pre test everything as a mini version before running RunAllTests

If this reports everything succesfully then we can revert the new line changes and proceed with RunAllTests

Rd4dev committed:
Reverting all new line changes after verifying the report

Rd4dev committed:
Reverting all new line changes after verifying the report 2

Rd4dev committed:
Triggering RunAllTests CI run

Rd4dev committed:
The previous run all tests failed since the env list was too big, to handle that the list is saved to a txt and passed to process. This commit sets compute all files to default false to just test if the changes work well also test exemption is added to run coverage just to speed up the check

Rd4dev committed:
Removing the dependency on the unit test workflow to run and verify the coverage workflow faster

Rd4dev committed:
Fixing bad substitution of pb_files.txt file in bash command

Rd4dev committed:
Reverting faster coverage report changes to set ready for RunAllTests

Rd4dev commented:
@BenHenning, also just found another issue with the existing implementation, since the CoverageReporter was moved to a new package, it was not computed properly, that needs to be fixed. Will do that once RunAllTests complete. 

Edit:
One thing I suspect, would performing any merge with other branches affect this? - Not really sure why.
But while fixing ambiguity error, I pushed a fix, then reverted to a previous commit, then switched to a new branch and merged changes from there, would these cause any issues, is this even allowed? 

I think that messed the commit history earlier and I doubt if I can proceed with this PR, after RunAllTests can I just create a new PR so this doesn't affect anything else.

Rd4dev committed:
New line change to test locally if changes to coverage reporter is making any occurence now

Rd4dev committed:
New line change to consolelogger as coveragereporter doesn't seem to report anything

Rd4dev committed:
Parsing txt file right in script instead of reading it in the workflow as that too causes arg list error

Rd4dev committed:
Remvoing suffix from the pb file path list as they don't seem to process the file in ci (though it did locally)

Rd4dev committed:
Including the moved file and trying to figure out ways to parse the list of pb file

Rd4dev committed:
Testing out the proto save state

Rd4dev committed:
Pre test push after fixing the list of pb file collection

If this works then add the compute all setting, add job dependency, remove test exemption

Rd4dev committed:
Running for a RunAllTests; but the git client test for move files is yet to be added

Running this in meantime I can figure other stuff out

Rd4dev committed:
Running RunAllTests without unit test dependency

Just realizing I don't need to wait for unit test to get the report for now

Rd4dev committed:
Test file exemption added with files that are overriding a min coverage percentage threshold unsorted

Rd4dev committed:
File exemption with overridden min coverage threshold sorted

Rd4dev committed:
Clean up after sorting the test exemptions; test to be added for move files

Rd4dev committed:
Added and updated tests with changes to retrieving renamed / moved files

Rd4dev committed:
Reverting commented unit test workflow dependency

Rd4dev committed:
Updated Kdocs for CoverageReporter with changes to providing proto paths as list.txt file

Rd4dev commented:
@BenHenning, the PR was tested with [RunAllTests] and the initial report was generated:  https://github.com/oppia/oppia-android/pull/5469#issuecomment-2286075490, with this the test exemptions to introduce overriding coverage percentages for all the files that failed below the min threshold are added to the test_file_exemptions.textproto file and ran again with [RunAllTests] to successfully Pass all coverage checks: https://github.com/oppia/oppia-android/pull/5469#issuecomment-2286506104.

**Additional fixes:** 
- The previously mentioned `main` ambiguity error was resolved by restructuring the file that had another main function in the same package to its sub package.

- CoverageReporter rename / move issue mentioned was resolved by taking into account all the renamed files. And their related test files are also added to the `GitClientTest.kt`. But I would like to check if this test case - `testUtility_featureBranch_movedFile_stagedAndCommitted_returnsNewFile` in the `computeChangedFilesTest` was functioning right? with the added functionality `moveFile`, since it did pass even without the new change to include the move files in the computed files. (But it failed with staging the oldFilePath, but with the new changes it accepts staging oldFilePath). Would like to confirm if the test case is functioning right.

- Argument list too long issue - was resolved by moving away from saving the pb_file list in ENVs to their own file and passing the file.txt to the script, which is then parsed in the script. (Added test cases for the same).

- Updated the template based on the M2 demo.

Can you PTAL?

BenHenning reviewed (approved):
Thanks @Rd4dev! I had some follow-up nits, but decided to drop them since they are minor and I'd rather see this get merged.

It is _fantastic_ to see the CI run working, and I'm amused to see how many shards 'RunAllTests' produces now.

The main thing left is to update the template to link to the new wiki article, but perhaps that is better to do in the wiki PR itself since it won't exist until then (plus we can them prioritize getting this merged).

BenHenning commented:
Holding off merge until CI is green (can't auto-merge since the new workflows won't be marked as blocking until this PR is merged).

BenHenning commented:
It actually occurs to me that this _could_ be the perfect moment to verify the new final "check code coverage results" workflow passes when expected, so I decided to mark it as blocking and enable auto-merge. If it works, the PR will be merged. :)
